node/src/chain.rs

```diff
@@ -528,20 +366,272 @@ pub async fn create_ethereum_networks_for_chain(
             ),
             web3.limit_for(&config.node),
         );
+
+        if call_only {
+            call_only_adapters.push(adapter);
+        } else {
+            adapters.push(adapter);
+        }
     }
 
-    parsed_networks.sort();
-    Ok(parsed_networks)
+    adapters.sort_by(|a, b| {
+        a.capabilities
+            .partial_cmp(&b.capabilities)
+            // We can't define a total ordering over node capabilities,
+            // so incomparable items are considered equal and end up
+            // near each other.
+            .unwrap_or(Ordering::Equal)
+    });
+
+    Ok(AdapterConfiguration::Rpc(EthAdapterConfig {
+        chain_id: network_name.into(),
+        adapters,
+        call_only: call_only_adapters,
+        polling_interval: Some(chain.polling_interval),
+    }))
+}
+
+pub async fn networks_as_chains(
+    config: &Arc<EnvVars>,
+    blockchain_map: &mut BlockchainMap,
+    node_id: &NodeId,
+    logger: &Logger,
+    networks: &Networks,
+    store: Arc<BlockStore>,
+    logger_factory: &LoggerFactory,
+    metrics_registry: Arc<MetricsRegistry>,
+    chain_head_update_listener: Arc<ChainHeadUpdateListener>,
+) {
+    let adapters = networks
+        .adapters
+        .iter()
+        .group_by(|a| a.chain_id())
+        .into_iter()
+        .map(|(chain_id, adapters)| (chain_id, adapters.into_iter().collect_vec()))
+        .collect_vec();
+
+    let substreams: Vec<&FirehoseAdapterConfig> = networks
+        .adapters
+        .iter()
+        .flat_map(|a| a.as_substreams())
+        .collect();
+
+    let chains = adapters.into_iter().map(|(chain_id, adapters)| {
+        let adapters: Vec<&AdapterConfiguration> = adapters.into_iter().collect();
+        let kind = adapters
+            .iter()
+            .map(|a| a.kind())
+            .reduce(|a1, a2| match (a1, a2) {
+                (BlockchainKind::Substreams, k) => k,
+                (k, BlockchainKind::Substreams) => k,
+                (k, _) => k,
+            })
+            .expect("each chain should have at least 1 adapter");
```

Comment:
It would be better to log an error and ignore that chain instead of aborting startup of the entire process.

Reply:
If this is a valid state of configuration where an empty section can be provided then sure, I would also like to know what we should do with that setup. 

Do you have a suggestion for the error message?

Reply:
If reading the config file validates that there is at least one provider, logs an error and removes the chain from the config otherwise, it would be ok to leave the `expect` there to say something like `validation should have checked we have at least one provider`. Otherwise, the error should be something like `Chain {name} does not have any providers configured. Ignoring it`

Reply:
updated